h/views/api/group_members.py

```diff
@@ -99,9 +100,22 @@ def add_member(context: GroupMembershipContext, request):
     if context.user.authority != context.group.authority:
         raise HTTPNotFound()
 
+    if request.body:
+        appstruct = EditGroupMembershipAPISchema().validate(json_payload(request))
+        roles = appstruct["roles"]
+    else:
+        # This doesn't mean the membership will be created with no roles:
+        # default roles will be applied by the service.
+        roles = None
+
     group_members_service = request.find_service(name="group_members")
 
-    membership = group_members_service.member_join(context.group, context.user.userid)
+    try:
+        membership = group_members_service.member_join(
+            context.group, context.user.userid, roles
+        )
+    except ConflictError as err:
+        raise HTTPConflict(str(err)) from err
```

Comment:
Return a 409 if there's an existing membership with different roles.

For backwards-compatibility if there's an existing membership with the _same_ roles it'll still return 200 OK.